stepper/stepper.c

```diff
@@ -0,0 +1,129 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "stepper.h"
+#include "../opto-fork/opto.h"
+
+#define A_GPIO 2
+#define B_GPIO 3
+#define C_GPIO 6
+#define D_GPIO 13
+#define COIL_COUNT 4
+
+void init_stepper() {
+    static const int stepper_pin[COIL_COUNT] = {A_GPIO,
+                                                B_GPIO,
+                                                C_GPIO,
+                                                D_GPIO};
+
+    for (int stepper_i = 0; stepper_i < COIL_COUNT; stepper_i++) {
+        gpio_init(stepper_pin[stepper_i]);
+        gpio_set_dir(stepper_pin[stepper_i], GPIO_OUT);
+    }
+}
+
+enum step_stage {
+    A  = 0b0001,
+    AB = 0b0011,
+    B  = 0b0010,
+    BC = 0b0110,
+    C  = 0b0100,
+    CD = 0b1100,
+    D  = 0b1000,
+    DA = 0b1001
+};
+
+#define STEP_STATES 8
+
+void step(bool clockwise) {
+    static const uint8_t step_masks[STEP_STATES] =
+            {A, AB, B, BC, C, CD, D, DA};
+
+    static const struct coil_struct {
+        int gpio;
+        uint8_t bit;
+    } coils[COIL_COUNT] = {{A_GPIO, A},
+                           {B_GPIO, B},
+                           {C_GPIO, C},
+                           {D_GPIO, D}};
+
+    static int8_t stepper_mask_i = 0;
+
+    if (clockwise) {
+        if (++stepper_mask_i > 7) stepper_mask_i = 0;
+    } else {
+        if (--stepper_mask_i < 0) stepper_mask_i = 7;
+    }
+
+    uint8_t step_state = step_masks[stepper_mask_i];
+
+    for (int coil_i = 0; coil_i < COIL_COUNT; coil_i++) {
+        struct coil_struct coil = coils[coil_i];
+        gpio_put(coil.gpio, coil.bit & step_state);
+    }
+}
+
+// Max stepper speed the disk can keep up with (according to Roni's experience)
+#define SPD_REDUC_MIN 800
+// Initial speed, for helping the disk accumulate speed gradually,
+// helping it build some momentum
+#define SPD_REDUC_INIT 1200
+
+#define CALIB_OPERATIONS 4
+#define OPTO_OFFSET 158
+
+int calibrate() {
+    uint64_t step_shift = SPD_REDUC_INIT;
+
+    printf("Calibrating...\n");
+
+    const bool operation_order[CALIB_OPERATIONS] = {false,
+                                                    true,
+                                                    false,
+                                                    true};
+    int steps = 0;
+
+    for (int oper_i = 0; oper_i < CALIB_OPERATIONS; oper_i++) {
```

Comment:
It's iterating through 4 phases:
1. Rotate **out** (hence 'false') of opto-fork. Because we don't know at which specific 'step' it is sitting on **if** it is in "the opto-fork" at all.
2. Search for the opto-fork (hence 'true'), confirming the very edge of the it.
3. Having found the opto-fork (and having accelerated to full speed), it can start counting the steps, first rotating **out** of the opto-fork...
4. ... and then rotate until it finds the opto-fork, again, counting the steps up until that point.

Like, I commented on the previous issue, I think it could need a face lift.